server/sonar-server/src/main/java/org/sonar/server/batch/ProjectDataLoader.java

```diff
@@ -0,0 +1,252 @@
+import org.sonar.core.permission.GlobalPermissions;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.MyBatis;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.db.component.FilePathWithHashDto;
+import org.sonar.db.property.PropertyDto;
+import org.sonar.server.exceptions.ForbiddenException;
+import org.sonar.server.user.UserSession;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+import static org.sonar.server.ws.WsUtils.checkFound;
+
+@ServerSide
+public class ProjectDataLoader {
+
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  public ProjectDataLoader(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+  }
+
+  public ProjectRepositories load(ProjectDataQuery query) {
+    boolean hasScanPerm = userSession.hasGlobalPermission(GlobalPermissions.SCAN_EXECUTION);
+    checkPermission(query.isIssuesMode());
+
+    DbSession session = dbClient.openSession(false);
+    try {
+      ProjectRepositories data = new ProjectRepositories();
+      ComponentDto module = checkFound(dbClient.componentDao().selectByKey(session, query.getModuleKey()),
+        format("Project or module with key '%s' is not found", query.getModuleKey()));
+
+      // Scan permission is enough to analyze all projects but preview permission is limited to projects user can access
+      if (query.isIssuesMode() && !userSession.hasProjectPermissionByUuid(UserRole.USER, module.projectUuid())) {
+        throw new ForbiddenException("You're not authorized to access to project '" + module.name() + "', please contact your SonarQube administrator.");
+      }
+
+      ComponentDto project = getProject(module, session);
+      if (!project.key().equals(module.key())) {
+        addSettings(data, module.getKey(), getSettingsFromParents(module, hasScanPerm, session));
+      }
+
+      List<ComponentDto> modulesTree = dbClient.componentDao().selectEnabledDescendantModules(session, module.uuid());
+      Map<String, String> moduleUuidsByKey = moduleUuidsByKey(modulesTree);
+      Map<String, Long> moduleIdsByKey = moduleIdsByKey(modulesTree);
+
+      List<PropertyDto> modulesTreeSettings = dbClient.propertiesDao().selectEnabledDescendantModuleProperties(module.uuid(), session);
+      TreeModuleSettings treeModuleSettings = new TreeModuleSettings(moduleUuidsByKey, moduleIdsByKey, modulesTree, modulesTreeSettings);
+
+      addSettingsToChildrenModules(data, query.getModuleKey(), Maps.<String, String>newHashMap(), treeModuleSettings, hasScanPerm);
+      List<FilePathWithHashDto> files = searchFilesWithHashAndRevision(session, module);
+      addFileData(data, modulesTree, files);
+
+      // FIXME need real value but actually only used to know if there is a previous analysis in local issue tracking mode so any value is
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)